server: add option to output probabilities for completion

WangHaoranRobin opened the pull request:
related to #1784 

I was playing around with the [server example](https://github.com/ggerganov/llama.cpp/blob/98ed16557432d7a5179c57eddcc3a08a7ae6d54d/examples/server/server.cpp) and wanted to expose the probabilities of the generated tokens to the server client to implement custom stopping sequences and criteria(similar to openai's api [here](https://platform.openai.com/docs/api-reference/completions/create#completions/create-logprobs)).

This PR returns the probabilities by reading it from `candidates_p.data[i].p` after sampling functions are called.

WangHaoranRobin committed:
server: add option to output probabilities for completion

WangHaoranRobin committed:
Merge pull request #1 from WangHaoranRobin/robin_fork_master

server: add option to output probabilities for completion

SlyEcho commented on examples/common.h:
```diff
@@ -31,6 +31,7 @@ struct gpt_params {
     int32_t main_gpu                        = 0;   // the GPU that is used for scratch and small tensors
     float   tensor_split[LLAMA_MAX_DEVICES] = {0}; // how split tensors should be distributed across GPUs
     bool    low_vram                        = 0;   // if true, reduce VRAM usage at the cost of performance
+    int32_t n_probs                         = 0;   // if greater than 1, output the probabilities of top n_probs tokens. Max 5
```
Why 5?

WangHaoranRobin replied:
I was copying openai's api. I think it might be trying to prevent people from doing a `n_probs = 10000` and taking up ram on the server? 

But max 5 is not implemented in common.h but in server.cpp. I will correct the comments.

WangHaoranRobin committed:
server: fix comment about max n_probs

WangHaoranRobin committed:
Merge pull request #2 from WangHaoranRobin/robin_fork_master

server: fix comment about max n_probs

SlyEcho commented on examples/server/server.cpp:
```diff
@@ -26,6 +26,28 @@ struct server_params {
     int32_t write_timeout = 600;
 };
 
+// completion string output with probabilities 
+struct completion_string_output {
+    struct token_prob {
+        std::string tok_str;
+        float prob;
+    };
+
+    std::vector<token_prob> probs;
+    std::string tok_str;
+};
```
Is it necessary to copy the strings? Maybe `completion_token_output` is all that is needed, it is very easy to go from token to string.

WangHaoranRobin replied:
Yes, I removed it in the newest commit.

SlyEcho commented on examples/server/server.cpp:
```diff
@@ -377,12 +406,22 @@ struct llama_server_context {
 
+        // iterate through token_with_probs.probs, if tok is valid, convert it to string and add to result.prob
+        for (const auto & prob : token_with_probs.probs) {
+            const std::string prob_text = prob.tok == -1 ? "" : llama_token_to_str(ctx, prob.tok);
```
It is not checking for validity? It will crash the server when a partial UTF-8 character is generated?

Maybe it should happen after the multibyte stuff below?

WangHaoranRobin replied:
My bad. I didn't take care of partial character generation at all in my first commit. Just added support in the newest commit.

> It would only apply if the other probable tokens were.

I now bundle the completion_token_output objects when dealing with multibyte. ie. if the character is 2 tokens long, we will return an array of 2 completion_token_output objects.

> I would prefer if all the code for token probabilities would be behind a flag in the request so it doesn't add overhead when is not needed.

Agree. Implemented this in the newest commit.

SlyEcho commented on examples/common.h:
```diff
@@ -31,6 +31,7 @@ struct gpt_params {
     int32_t main_gpu                        = 0;   // the GPU that is used for scratch and small tensors
     float   tensor_split[LLAMA_MAX_DEVICES] = {0}; // how split tensors should be distributed across GPUs
     bool    low_vram                        = 0;   // if true, reduce VRAM usage at the cost of performance
+    int32_t n_probs                         = 0;   // if greater than 0, output the probabilities of top n_probs tokens.
```
It should probably not even be in common.h. The server has its own parameter struct in server.cpp.

WangHaoranRobin replied:
I put it in common.h because I thought other examples might use n_probs for custom stop sequences too. Like a chatbot.

WangHaoranRobin committed:
server: fix issue when handling probability output for incomplete tokens for multibyte character generation

WangHaoranRobin committed:
Merge pull request #3 from WangHaoranRobin/robin_fork_master

server: fix issue for multibyte character generation

WangHaoranRobin committed:
server: fix some beginner mistakes

WangHaoranRobin committed:
Merge pull request #4 from WangHaoranRobin/robin_fork_master

server: fix some beginner mistakes

WangHaoranRobin commented:
I have a question about partial UTF-8 character is generation:

In `doCompletion`:
https://github.com/ggerganov/llama.cpp/blob/7487137227eb32ed9b12156338b865cb29b2dfd1/examples/server/server.cpp#L388-L402

It seems like we are assuming that all partial character generation is 1 char long when converted to string. Is that true? I kinda used this assumption in [tokens_to_output_formatted_string](https://github.com/WangHaoranRobin/llama.cpp/blob/7cd8fc20d0c9b1329c365b82689f3c984a7ef0aa/examples/server/server.cpp#L101)

WangHaoranRobin committed:
Merge branch 'ggerganov:master' into master

WangHaoranRobin committed:
server: remove trailling white space

WangHaoranRobin committed:
Merge pull request #5 from WangHaoranRobin/robin_fork_master

server: remove trailling white space

WangHaoranRobin committed:
Merge branch 'ggerganov:master' into master

WangHaoranRobin committed:
Merge branch 'ggerganov:master' into master

WangHaoranRobin commented:
@SlyEcho are there any other requested changes? Should we merge this?

SlyEcho commented on examples/server/server.cpp:
```diff
@@ -312,35 +361,38 @@ struct llama_server_context {
+                    result.tok = llama_sample_token(ctx, &candidates_p);
                 }
             }
+            // Add maximum of 5 most probable tokens to the result
+            for (size_t i = 0; i < std::min(candidates_p.size, std::min((size_t) n_probs, size_t(5))); ++i) {
```
```suggestion
            for (size_t i = 0; i < std::min(candidates_p.size, (size_t) n_probs); ++i) {
```

SlyEcho commented:
I'm taking a look now. I was occupied on the weekend by holidays and a certain bald Russian clown.

1. Pretty sure any limit on the number of probabilities is unnecessary. OpenAI may have some incentive to limit their API, but for a user's own machine it's insignificant.
2. The Unicode handling is just wrong, IMO. We already detect a partial mutliple byte character and in that situation it should just add to the probabilites list of the last token somehow.

WangHaoranRobin commented:
> by holidays and a certain bald Russian clown

LOL


What do you mean by 
> add to the probabilities list of the last token?

Are you saying there should be something is the server context like `std::vector<completion_token_output> current_token_probs;` ? Isn't that just repeated information with `generated_token_probs`?

I feel like [the current logic](https://github.com/WangHaoranRobin/llama.cpp/blob/af058cf82083892536a9b23266178967cf2ba556/examples/server/server.cpp#L959-L967) is the easiest. Under the assumption that `llama.generated_text` to `llama.generated_token_probs` is bijective, i.e. we send the same amount of probs object as tokens.

SlyEcho commented:
I'm currently thinking what would be best way. It's also possible that other alternative tokens could come out as incomplete and should we keep generating until we have `n_probs` amount of complete items?

SlyEcho commented on examples/server/server.cpp:
```diff
@@ -86,6 +97,40 @@ static void server_log(const char * level, const char * function, int line,
     fflush(stdout);
 }
 
+// format incomplete utf-8 multibyte character for output
+static std::string tokens_to_output_formatted_string(const llama_context * ctx, const llama_token token) {
+    std::string out = token == -1 ? "" : llama_token_to_str(ctx, token);
+    // if first bit is 1, meaning it's a partial character
+    if ((out[0] & 0x80) == 0x80) {
+        std::stringstream ss;
+        ss<< std::hex << (out[0] & 0xff);
+        std::string res ( ss.str() );
+        out = "byte: \\x" + res;
+    }
+    return out;
+}
```
This function seems kind of hacky. What format is `byte: \x00` supposed to be?

It also seems incorrect as it only looks at the first character, but that doesn't mean that the continuation bytes are not there.

Maybe for the first version we can just let the JSON library replace invalid sequences with �

WangHaoranRobin replied:
based on https://huggingface.co/oobabooga/llama-tokenizer/raw/main/tokenizer.json anon provided, all incomplete tokens are one byte long. So if `out` is a partial char, it should be exactly one byte and starts with 0b1.

WangHaoranRobin replied:
The `byte: \x00` is for easy reading and printing on the client side. Based on the syntax [here](https://design215.com/toolbox/ascii-utf8.php#:~:text=UTF%2D8%202%2Dbyte%20Characters%3A%20byte%201%20%3D%20%5Cxc0%2D%5Cxdf%2C%20byte%202%20%3D%20%5Cx80%2D%5Cxbf)

SlyEcho replied:
Well, it's not easy because it's supposed to be machine readable. Maybe an option is to encode it in a different JSON property as a number value?

WangHaoranRobin commented:
> should we keep generating until we have `n_probs` amount of complete items?

As in that we should continue generation for all `n_probs` possible scenarios we got from last inference? If we are talking about 4 byte characters and `n_probs = 5`, that's 5^4=625 possible complete utf-8 characters.

WangHaoranRobin committed:
server: remove n_probs upper limit of 5

WangHaoranRobin committed:
Merge pull request #6 from WangHaoranRobin/robin_fork_master

server: remove n_probs upper limit of 5

SlyEcho commented:
> As in that we should continue generation for all n_probs possible scenarios we got from last inference? If we are talking about 4 byte characters and n_probs = 5, that's 5^4=625 possible complete utf-8 characters.

Yeah, this is kind of a lot, we could keep the best 5 but it would mean a lot of token generating. Or to only keep the alternates from the last one.

WangHaoranRobin commented:
Not that we need to copy exactly openai's API. But here is an example of how they handle multibyte:
>"top_logprobs": [
          {
            "bytes:\\xef": -0.5548667,
            "\n": -0.9359168,
            "?": -3.5638626,
            "bytes:\\xe5\\x91": -6.030914,
            "bytes:\\xe5": -6.4047127
          },
          {
            "bytes:\\xbc": -2.2159093e-06,
            "bytes:\\xbd": -13.45378,
            "bytes:\\xbb": -14.923321,
            "bytes:\\xb9": -15.295079,
            "bytes:\\xbf": -16.138496
          },
          {
            "bytes:\\x9f": -0.0011532516,
            "bytes:\\x8c": -7.17465,
            "bytes:\\x9a": -7.992851,
            "bytes:\\x81": -10.874427,
            "bytes:\\x89": -11.387907
          },
...

And I think we shouldn't just return probs of the last token as the probs of other partial chars are also useful information.

> we could keep the best 5 but it would mean a lot of token generating

This would mean `n_probs` times inference time(50ms -> 250ms at `n_probs`=5, when the user only intended to get information about token probs). Which I don't find reasonable.

SlyEcho commented:
Ok, let's leave it like that for now and try to think how to improve in the future.

SlyEcho commented on examples/server/server.cpp:
```diff
@@ -312,35 +361,38 @@ struct llama_server_context {
+                    result.tok = llama_sample_token_mirostat_v2(ctx, &candidates_p, mirostat_tau, mirostat_eta, &mirostat_mu);
                 } else {
                     // Temperature sampling
                     llama_sample_top_k(ctx, &candidates_p, top_k, 1);
```
I think this should change, otherwise it doesn't return the requested amount of items?

```suggestion
                    llama_sample_top_k(ctx, &candidates_p, top_k, std::max(1, n_probs));
```

WangHaoranRobin replied:
Yes my bad. Updated in the newest commit

WangHaoranRobin committed:
server: handle probs output when temp=0; handle final response probs output

WangHaoranRobin committed:
Merge branch 'master' into robin_fork_master

WangHaoranRobin committed:
Merge pull request #7 from WangHaoranRobin/robin_fork_master

server: handle probs output when temp=0; handle final response probs output

WangHaoranRobin commented:
I added an extra llama_sample_softmax when temp=0, i.e. when we use greedy sampling, else the candidate list is not sorted and data.p are not assigned. 

I also changed so that the final response of streamed results contains all the probs.

SlyEcho commented:
The evaluated prompt should also return probabilities or not?

WangHaoranRobin commented:
I dont think so. 
1. for batch size > 1, we don't have probs for every token. and 
2. From a server client perspective, I don't think the implied affordance here is to include the probs of input prompt.

WangHaoranRobin committed:
Merge branch 'ggerganov:master' into master

SlyEcho commented:
I can think of a few uses for it but maybe it can be added later.

SlyEcho reviewed (approved):
OK, it looks good enough now.

WangHaoranRobin committed:
Merge branch 'ggerganov:master' into master

WangHaoranRobin committed:
server: fix llama_sample_top_k order

WangHaoranRobin committed:
Merge pull request #8 from WangHaoranRobin/robin_fork_master

server: fix llama_sample_top_k order

WangHaoranRobin commented on examples/server/server.cpp:
```diff
@@ -312,35 +361,42 @@ struct llama_server_context {
+                    llama_sample_tail_free(ctx, &candidates_p, tfs_z, min_keep);
+                    llama_sample_typical(ctx, &candidates_p, typical_p, min_keep);
+                    llama_sample_top_p(ctx, &candidates_p, top_p, min_keep);
+                    llama_sample_top_k(ctx, &candidates_p, top_k, min_keep);
```
thanks! 👌🏻 fixed

WangHaoranRobin committed:
Merge branch 'ggerganov:master' into master

ggerganov commented on examples/common.h:
```diff
@@ -32,6 +32,7 @@ struct gpt_params {
     int32_t main_gpu                        = 0;   // the GPU that is used for scratch and small tensors
     float   tensor_split[LLAMA_MAX_DEVICES] = {0}; // how split tensors should be distributed across GPUs
     bool    low_vram                        = 0;   // if true, reduce VRAM usage at the cost of performance
```
Minor nit : this `bool` should be together with the other `bool`s in the struct

WangHaoranRobin replied:
Just changed. It doesn;t really fit in the topic of this pr tho.

WangHaoranRobin committed:
Merge branch 'ggerganov:master' into master

WangHaoranRobin committed:
examples/common.h: put all bool variables in gpt_params together

WangHaoranRobin committed:
examples/common.h: put all bool variables in gpt_params together

WangHaoranRobin committed:
Merge pull request #9 from WangHaoranRobin/robin_fork_master

examples/common.h: put all bool variables in gpt_params together
